python/pylibcudf/pylibcudf/expressions.pyx

```diff
@@ -251,3 +258,190 @@ cdef class ColumnNameReference(Expression):
+    """
+
+    def __init__(self, dict column_mapping):
+        self.column_mapping = column_mapping
+
+    def visit_Name(self, node):
+        try:
+            return self.column_mapping[node.id]
+        except KeyError:
+            raise ValueError(f"Unknown column name {node.id}")
+
+    def visit_Constant(self, node):
+        if not isinstance(node.value, (float, int, str, complex)):
+            raise ValueError(
+                f"Unsupported literal {repr(node.value)} of type "
+                "{type(node.value).__name__}"
+            )
+        return Literal(from_arrow(pa.scalar(node.value)))
+
+    def visit_UnaryOp(self, node):
+        operand = self.visit(node.operand)
+        if isinstance(node.op, ast.USub):
+            # TODO: Except for leaf nodes, we won't know the type of the
+            # operand, so there's no way to know whether this should be a float
+            # or an int. We should maybe see what Spark does, and this will
+            # probably require casting.
+            minus_one = Literal(from_arrow(pa.scalar(-1)))
+            return Operation(ASTOperator.MUL, minus_one, operand)
+        elif isinstance(node.op, ast.UAdd):
+            return operand
+        else:
+            op = _python_cudf_operator_map[type(node.op)]
+            return Operation(op, operand)
+
+    def visit_BinOp(self, node):
+        left = self.visit(node.left)
+        right = self.visit(node.right)
+        op = _python_cudf_operator_map[type(node.op)]
+        return Operation(op, left, right)
+
+    def visit_BoolOp(self, node):
+        return functools.reduce(
+            functools.partial(Operation, ASTOperator.LOGICAL_AND),
+            (
+                Operation(node.op, self.visit(left), self.visit(right))
+                for left, right in zip(
+                    node.values[:-1], node.values[1:], strict=True
+                )
+            )
+        )
+
+    def visit_Compare(self, node):
+        operands = [node.left, *node.comparators]
+        return functools.reduce(
+            functools.partial(Operation, ASTOperator.LOGICAL_AND),
+            (
+                Operation(op, self.visit(left), self.visit(right))
+                for op, left, right in zip(
+                    node.ops, operands[:-1], operands[1:], strict=True
+                )
+            )
+        )
+
+    def visit_Call(self, node):
+        try:
+            op = _python_cudf_function_map[node.func.id]
+        except KeyError:
+            raise ValueError(f"Unsupported function {node.func}.")
+        # Assuming only unary functions are supported, which is checked above.
+        if len(node.args) != 1 or node.keywords:
+            raise ValueError(
+                f"Function {node.func} only accepts one positional "
+                "argument."
+            )
+        return Operation(op, self.visit(node.args[0]))
+
+
+@functools.lru_cache(256)
+def to_expression(str expr, dict column_mapping):
```

Comment:
Ah, if we want this to be a cache, then we can't use a dict here.

Also, I think that `compute_column` doesn't work with `ColumnNameReference`, only `ColumnReference`. So let us construct things internally:

```suggestion
@functools.lru_cache(256)
def to_expression(str expr, tuple column_names):
```

We can handle the parquet filter case in a second round.